tests/data_manager.rs

```diff
@@ -689,6 +690,280 @@ fn mutable_data_parallel_mutations() {
+                            }) {
+                    expected_successes = 1;
+                }
+
+                trace!("Updating data {:?} with actions {:?}", data.name(), actions);
+                let msg_id =
+                    clients[0].mutate_mdata_entries(*data.name(), data.tag(), actions.clone());
+                let _ = sent_actions.insert(msg_id, actions);
+            }
+        }
+
+        event_count += poll::nodes_and_clients_parallel(&mut nodes, &mut clients);
+        trace!("Processed {} events.", event_count);
+
+        let mut successes: usize = 0;
+
+        while let Ok(event) = clients[0].try_recv() {
+            if let Event::Response {
+                       response: Response::MutateMDataEntries { res, msg_id }, ..
+                   } = event {
+                match res {
+                    Ok(()) => {
+                        trace!("Client {:?} received successful response.",
+                               clients[0].name());
+                        let actions = unwrap!(sent_actions.remove(&msg_id));
+                        unwrap!(all_data[index].mutate_entries(actions,
+                                                               *clients[0].signing_public_key()));
+                        successes += 1;
+                    }
+                    Err(error) => {
+                        trace!("Client {:?} received failed response. Reason: {:?}",
+                               clients[0].name(),
+                               error);
+                    }
+                }
+            }
+        }
+
+        assert_eq!(successes, expected_successes);
+        mock_crust_detail::check_data(all_data.iter().cloned().map(Data::Mutable).collect(),
+                                      &nodes);
+
+        let sorted_nodes = test_node::closest_to(&nodes, clients[0].name(), GROUP_SIZE);
+        let node_count_stats: Vec<_> = sorted_nodes
+            .into_iter()
+            .map(|node| (node.name(), node.get_maid_manager_mutation_count(clients[0].name())))
+            .collect();
+        expected_mutation_count += successes;
+        for &(_, count) in &node_count_stats {
+            assert_eq!(Some(expected_mutation_count as u64),
+                       count,
+                       "Expected {} mutations got: {:?}",
+                       expected_mutation_count,
+                       node_count_stats);
+        }
+    }
+
+    mock_crust_detail::verify_network_invariant_for_all_nodes(&nodes);
+    // Check that the stored data matches the local copy.
+    verify_data_is_stored(&mut nodes, &mut clients[0], &all_data);
+}
+
+// Two clients concurrently mutating same data with same actions, one with permission, the other not.
+// Only the one with permission shall succeed.
+#[test]
+fn no_permission_mutable_data_concurrent_mutations() {
+    let seed = None;
+    let node_count = TEST_NET_SIZE;
+    let client_count = 2;
```

Comment:
Minor: I don't think we need to extract the client count here, as this test requires precisely two clients. I'd say it's fine to hardcode it on line 849.